File: DictionaryProject/Commands/RandomCommand.vb
<Highlight(RegEx:="^\s*random \\\{", Color:="Green", NextLevel:=3)> _
<Highlight(Regex:="^\s*\d+(\.\d+)?", Color:="Red", Level:=3)> _
<Highlight(Regex:="\\\}", Color:="Green", Level:=3, NextLevel:=-2)> _
Public Class RandomCommand
    Inherits CommandBase

    Public Overrides Sub CheckSanity(ByVal project As Project)
        For Each cmd As CommandBase In _Commands.Select(Function(w As WeightedCommand) w.Command)
            cmd.CheckSanity(project)
        Next
    End Sub

    Private Shared _Random As New Random()

    Public Overrides Sub Execute(ByVal context As Context)
        Dim total As Double = Commands.Sum(Function(wc As WeightedCommand) wc.Weight)

        Dim selected As Double = _Random.NextDouble() * total

        For Each wc As WeightedCommand In Commands
            selected -= wc.Weight
            If selected <= 0 Then
                wc.Command.Execute(context)
                Exit Sub
            End If
        Next
    End Sub

    Private _Commands As New List(Of WeightedCommand)
    Public ReadOnly Property Commands() As List(Of WeightedCommand)
        Get
            Return _Commands
        End Get
    End Property

    Public Overrides Sub LoadCommand(ByVal project As Project, ByVal reader As System.IO.TextReader, ByVal line As String, ByRef lineNumber As Integer)
        MyBase.LoadCommand(project, reader, line, lineNumber)

        Dim parts As List(Of String) = ProjectSerializer.ReadTokens(line)

        If parts.Last <> "{" Then
            project.Warnings.Add(String.Format("Line {0}: The Random command requires a {{ at the end of the line.", lineNumber))
        Else
            parts.RemoveAt(0)
            parts.RemoveAt(parts.Count - 1)

            If parts.Count = 0 Then
                If ProjectSerializer.ReadLines(New Object() {project, Commands}, reader, AddressOf ProjectSerializer.ParseWeightedCommands, lineNumber) Then
                    project.Warnings.Add(String.Format("Line {0}: The Random command is not closed correctly.", lineNumber))
                End If
            Else
                project.Warnings.Add(String.Format("Line {0}: The Random command must have no arguments.", lineNumber))
            End If
        End If
    End Sub

    Public Overrides Sub WriteCommand(ByVal writer As System.IO.TextWriter)
        writer.WriteLine("Random {")

        For Each wc As WeightedCommand In _Commands
            writer.Write("{0} ", wc.Weight.ToString(System.Globalization.CultureInfo.InvariantCulture))
            wc.Command.WriteCommand(writer)
        Next

        writer.WriteLine("}")
    End Sub
End Class

File: monowordbuilder/wordbuilderbase/Commands/RandomCommand.cs
using System;
using System.Collections.Generic;

namespace Whee.WordBuilder.Model.Commands
{
	[Highlight(RegEx = "^\\s*random \\\\\\{", Color = "Green", NextLevel = 3)]
	[Highlight(RegEx = "^\\s*\\d+(\\.\\d+)?", Color = "Red", Level = 3)]
	[Highlight(RegEx = "\\\\\\}", Color = "Green", Level = 3, NextLevel = -2)]
	public class RandomCommand : CommandBase
	{
	
		private static Random _Random = new Random();
	
		public override void Execute(Context context)
		{
			double total = 0;
			foreach (WeightedCommand wc in Commands) 
			{
				total += wc.Weight;
			}
	
			double selected = _Random.NextDouble() * total;
	
			foreach (WeightedCommand wc in Commands) {
				selected -= wc.Weight;
				if (selected <= 0) {
					wc.Command.Execute(context);
					return;
				}
			}
		}
	
		private List<WeightedCommand> _Commands = new List<WeightedCommand>();
		public List<WeightedCommand> Commands {
			get { return _Commands; }
		}

        public override void LoadCommand(Whee.WordBuilder.ProjectV2.IProjectSerializer serializer)
        {
            ProjectV2.WeightedCommandBlockNode wcbn = new ProjectV2.WeightedCommandBlockNode(serializer);

            Children.Add(wcbn);

            foreach (ProjectV2.WeightedCommandNode wcn in wcbn.Children)
            {
                WeightedCommand wc = new WeightedCommand();
                wc.Weight = wcn.Weight;
                wc.Command = wcn.Command;
                Commands.Add(wc);
            }
        }
	
		public override void LoadCommand(Project project, System.IO.TextReader reader, string line, ref int lineNumber)
		{
			base.LoadCommand(project, reader, line, ref lineNumber);
	
			List<string> parts = ProjectSerializer.ReadTokens(line);
	
			if (parts[parts.Count - 1] != "{") {
				project.Warnings.Add(string.Format("Line {0}: The Random command requires a {{ at the end of the line.", lineNumber));
			}
			else {
				parts.RemoveAt(0);
				parts.RemoveAt(parts.Count - 1);
	
				if (parts.Count == 0) {
					if (ProjectSerializer.ReadLines(new object[] { project, Commands }, reader, ProjectSerializer.ParseWeightedCommands, ref lineNumber)) {
						project.Warnings.Add(string.Format("Line {0}: The Random command is not closed correctly.", lineNumber));
					}
				}
				else {
					project.Warnings.Add(string.Format("Line {0}: The Random command must have no arguments.", lineNumber));
				}
			}
		}
	
		public override void WriteCommand(System.IO.TextWriter writer)
		{
			writer.WriteLine("Random {");
	
			foreach (WeightedCommand wc in _Commands) {
				writer.Write("{0} ", wc.Weight.ToString(System.Globalization.CultureInfo.InvariantCulture));
				wc.Command.WriteCommand(writer);
			}
	
			writer.WriteLine("}");
		}

        public override void CheckSanity(Project project, Whee.WordBuilder.ProjectV2.IProjectSerializer serializer)
        {
            foreach (WeightedCommand cmd in _Commands)
            {
                cmd.Command.CheckSanity(project, serializer);
            }
        }

        public override bool RequireNewLine
        {
            get
            {
                return false;
            }
        }
    }
}